In [1]:
import pandas as pd 

1. En pairs anteriores hemos ido pudiendo sacar algunas conclusiones y contestando las preguntas que nos hacían las jefas. Desde la dirección se están planteando cambiar el sistema de asiganción de las solicitudes que llegan. En concreto, necesitan saber que categoría es la que recibe más tickets para poder asignar más empleados a esa categoría. Por esto nos piden que averiguemos qué categoría es la que tiene más tickets. Además nos piden que devolvamos esta consulta en un DataFrame.

📌 NOTA Ordena los resultados de mayor a menor. Esto todavía no lo hemos visto, pero un DataFrame lo podemos ordenar con .sort_values(). Aquí   tenéis una explicación de cómo hacerlo.

In [2]:
df_tickets = pd.read_csv("datos/tickets_years.csv")

In [8]:
df_tickets["categoría"].unique()

array(['Acceso&%Login', 'Sistema', 'Software', 'Hardware', nan,
       'Acceso%&Login', 'Acceso-&Login', 'Acceso99Login', 'Acceso/Login'],
      dtype=object)

In [11]:
df_tickets_categoria = df_tickets.groupby("categoría")["id_ticket"].count().reset_index()

In [44]:
df_tickets_categoria = df_tickets_categoria.sort_values(by = "id_ticket", ascending = False)

In [45]:
df_tickets_categoria

categoría  id_ticket
6        Sistema      38997
7       Software      19566
5       Hardware       9731
3   Acceso/Login       8659
4  Acceso99Login       6426
2  Acceso-&Login       5672
0  Acceso%&Login       4526
1  Acceso&%Login       3907

La categoría con más tickets es Sistema, con 38997

2. Desde la dirección quieren seguir evaluando la situación planteada en el ejercicio anterior. Pero ahora quieres saber cual es la media de la satisfacción por categoría. En concreto quieren saber si independientemente de la categoría los clientes salen satisfechos, para que en caso de que no sea así poder identificar un posible problema subyacente. Por lo tanto, nos piden que calculemos la media de satisfacción por cada una de las categorías que tenemos en el dataframe.

In [13]:
df_satisfaccion_categoria = df_tickets.groupby("categoría")["satisfacción"].mean().reset_index()
df_satisfaccion_categoria

categoría  satisfacción
0  Acceso%&Login      4.059876
1  Acceso&%Login      3.954662
2  Acceso-&Login      4.087447
3   Acceso/Login      4.172075
4  Acceso99Login      4.105198
5       Hardware      4.100925
6        Sistema      4.102303
7       Software      4.106329

La media de satisfacción es aproximadamente 4 para todas las categorías.

3. Quieren mejorar la consulta anterior y ahora nos piden que además de la media les devolvamos la desviación estándar de la satisfacción.

In [15]:
df_satisfaccion_stats = df_tickets.groupby("categoría")["satisfacción"].agg(media_de_satisfaccion = ("mean"), desviacion_estandar_satisfaccion = ("std")).reset_index()
df_satisfaccion_stats

categoría  media_de_satisfaccion  desviacion_estandar_satisfaccion
0  Acceso%&Login               4.059876                          1.292629
1  Acceso&%Login               3.954662                          1.344462
2  Acceso-&Login               4.087447                          1.272903
3   Acceso/Login               4.172075                          1.197885
4  Acceso99Login               4.105198                          1.256943
5       Hardware               4.100925                          1.262979
6        Sistema               4.102303                          1.253662
7       Software               4.106329                          1.254547

4. En nuestros datos, además hemos visto que no solo tenemos categorías, y que las diferencias entre ellas no son especialmente significativas. Revisando los datos de nuevo, desde dirección saben que hay dos tipos de tickets (Problema y Solicitud), basándonos en esto, ¿hay diferencias en la satisfacción basándonos en la categoría y el tipo de tickets?.

In [18]:
df_sat_categoria_tipo = df_tickets.groupby(["tipo", "categoría"])["satisfacción"].mean().reset_index()
df_sat_categoria_tipo

tipo      categoría  satisfacción
0    Problema  Acceso%&Login      3.992799
1    Problema  Acceso&%Login      3.987052
2    Problema  Acceso-&Login      4.045079
3    Problema   Acceso/Login      4.228160
4    Problema  Acceso99Login      4.090354
5    Problema       Hardware      4.089293
6    Problema        Sistema      4.098409
7    Problema       Software      4.111156
8   Solicitud  Acceso%&Login      4.081698
9   Solicitud  Acceso&%Login      3.943448
10  Solicitud  Acceso-&Login      4.100438
11  Solicitud   Acceso/Login      4.153527
12  Solicitud  Acceso99Login      4.110276
13  Solicitud       Hardware      4.104656
14  Solicitud        Sistema      4.103600
15  Solicitud       Software      4.104709

No vemos grandes diferencias en la media de satisfacción por categoría y tipo.

5. A pesar de que sabemos que tenemos muchos empleados y varios departamentos, nuestras jefas no saben cuántas personas hay en cada uno de estos departamentos. Por esto, nos piden que calculemos el número de empleados que tenemos por departamento. Devolved estos resultados en un dataframe. Ordenadad los resultados de mayor a menor basándonos en la cantidad de empleados que tenemos.

In [19]:
df_empleo = pd.read_csv("datos/empleo.csv")
df_empleo.head()

Unnamed: 0 turno  id_empleado    apellido    nombre id_puesto  \
0           0     A            6       VILLA     JESUS  50-32200   
1           1   NaN           60        DIAZ  PATRICIA  50-32200   
2           2     B            2      DUARTE   ANTONIO  50-32200   
3           3   NaN           24  VILLANUEVA     JESUS  50-32200   
4           4   NaN           98    GALLEGOS  PATRICIA  50-32200   

                             email                   puesto  \
0        jesus.villa@tractchun.com  Comprador de Materiales   
1      patricia.diaz@tractchun.com  Comprador de Materiales   
2     antonio.duarte@tractchun.com  Comprador de Materiales   
3   jesus.villanueva@tractchun.com  Comprador de Materiales   
4  patricia.gallegos@tractchun.com  Comprador de Materiales   

            departamento     tipo_puesto  
0  CONTROL DE PRODUCCION  Administrativo  
1  CONTROL DE PRODUCCION  Administrativo  
2  CONTROL DE PRODUCCION  Administrativo  
3  CONTROL DE PRODUCCION  Administrativo  
4  CONTROL DE PRODUCCION  Administrativo

In [25]:
df_empleados_dep = df_empleo.groupby("departamento")["id_empleado"].count().reset_index()
df_empleados_dep

departamento  id_empleado
0   AUDITORIA INTERNA, CALIBRACION Y METROLOGIA           70
1                                  CONTABILIDAD           25
2                         CONTROL DE PRODUCCION          127
3                         INGENIERIA DE CALIDAD          395
4                     INGENIERIA DE MANUFACTURA          228
5                                 MANTENIMIENTO          258
6                                    MATERIALES          125
7                                      PERSONAL           20
8                                    PRODUCCION          475
9                              Recursos Humanos          117
10                  SISTEMAS, TRAFICO Y ADUANAS          160

6. BONUS Quieren saber que departamentos son los que tienen de media un indice de satisfacción. 💡Pista Para responder a esta pregunta tendremos que unir los dataframes con métodos aprendidos en lecciones anteriores de Pandas.

In [26]:
display(df_tickets.head(1))
display(df_empleo.head(1))

Unnamed: 0          id_ticket       fecha  id_empleado  id_agente  \
0           0  GDDENR-5042564453  2016-07-13       1735.0        4.0   

       categoría      tipo           severidad        prioridad  \
0  Acceso&%Login  Problema  0 - Sin Clasificar  0 - Sin Asignar   

   días_resolución  satisfacción   año     semestre  
0              0.0           5.0  2016  2º semestre

Unnamed: 0 turno  id_empleado apellido nombre id_puesto  \
0           0     A            6    VILLA  JESUS  50-32200   

                       email                   puesto           departamento  \
0  jesus.villa@tractchun.com  Comprador de Materiales  CONTROL DE PRODUCCION   

      tipo_puesto  
0  Administrativo

In [27]:
df_completo = df_tickets.merge(df_empleo, on='id_empleado')
df_completo.head()

Unnamed: 0_x          id_ticket                fecha  id_empleado  \
0             0  GDDENR-5042564453           2016-07-13       1735.0   
1          3240  GWLESR-6142530250           2016-06-09       1735.0   
2          4858  KDLTSR-9842504751           2016-05-14       1735.0   
3          7035  KMLTSR-6042483650           2016-04-23       1735.0   
4           593  GDLTNR-0043045356  2017-11-06 00:00:00       1735.0   

   id_agente      categoría       tipo           severidad        prioridad  \
0        4.0  Acceso&%Login   Problema  0 - Sin Clasificar  0 - Sin Asignar   
1       32.0        Sistema   Problema          2 - Normal         1 - Baja   
2       27.0        Sistema  Solicitud          2 - Normal  0 - Sin Asignar   
3       26.0        Sistema  Solicitud          2 - Normal        2 - Media   
4       23.0  Acceso%&Login  Solicitud          2 - Normal  0 - Sin Asignar   

   días_resolución  ...     semestre  Unnamed: 0_y turno  apellido nombre  \
0              0.0  ...  2º semestre           796   NaN   PORCHAS  OCHOA   
1             11.0  ...  1º semestre           796   NaN   PORCHAS  OCHOA   
2              8.0  ...  1º semestre           796   NaN   PORCHAS  OCHOA   
3             10.0  ...  1º semestre           796   NaN   PORCHAS  OCHOA   
4              0.0  ...  2º semestre           796   NaN   PORCHAS  OCHOA   

  id_puesto                        email       puesto departamento  \
0  10-32110  ochoa.porchas@tractchun.com  Almacenista   MATERIALES   
1  10-32110  ochoa.porchas@tractchun.com  Almacenista   MATERIALES   
2  10-32110  ochoa.porchas@tractchun.com  Almacenista   MATERIALES   
3  10-32110  ochoa.porchas@tractchun.com  Almacenista   MATERIALES   
4  10-32110  ochoa.porchas@tractchun.com  Almacenista   MATERIALES   

      tipo_puesto  
0  Administrativo  
1  Administrativo  
2  Administrativo  
3  Administrativo  
4  Administrativo  

[5 rows x 22 columns]

In [28]:
df_sat_dep = df_completo.groupby("departamento")["satisfacción"].mean().reset_index()
df_sat_dep

departamento  satisfacción
0   AUDITORIA INTERNA, CALIBRACION Y METROLOGIA      4.075758
1                                  CONTABILIDAD      4.051587
2                         CONTROL DE PRODUCCION      4.085525
3                         INGENIERIA DE CALIDAD      4.109153
4                     INGENIERIA DE MANUFACTURA      4.095234
5                                 MANTENIMIENTO      4.087153
6                                    MATERIALES      4.125700
7                                      PERSONAL      4.097264
8                                    PRODUCCION      4.101763
9                              Recursos Humanos      4.121715
10                  SISTEMAS, TRAFICO Y ADUANAS      4.102721

7. BONUS Por último, quieres conocer los 5 empleados que tienen mayor valor en la satisfacción y los 5 empleados que tienen un menor valor de satisfacción. Todo esto respecto a la media. En este caso, conseguiremos una dataframe, donde tengamos el el id del empleado y su satisfacción, pero no sabemos su nombre y appelidos. ¿Se os ocurre como conseguir la información del nombre del empleado?

Para conseguir nombres y apellidos tenemos que unir los dataframes como en el ejercicio anterior.

In [30]:
df_completo.sample(4)

Unnamed: 0_x          id_ticket                fecha  id_empleado  \
1073           3868  GHLTSR-9743717794  2019-09-09 00:00:00        979.0   
91273         15477  KWLENR-0044049695  2020-08-06 00:00:00       1679.0   
44316          7514  KHLENR-8043332436  2018-08-20 00:00:00        783.0   
87250          8519  SDLTSR-9742647621           2016-10-04        552.0   

       id_agente      categoría       tipo   severidad        prioridad  \
1073        17.0        Sistema  Solicitud  2 - Normal         3 - Alta   
91273       36.0   Acceso/Login   Problema  2 - Normal         1 - Baja   
44316       34.0  Acceso-&Login   Problema  2 - Normal         3 - Alta   
87250       26.0        Sistema  Solicitud  2 - Normal  0 - Sin Asignar   

       días_resolución  ...     semestre  Unnamed: 0_y turno    apellido  \
1073               7.0  ...  2º semestre           889   NaN       AMAYA   
91273              0.0  ...  2º semestre           144   NaN  MONTENEGRO   
44316              0.0  ...  2º semestre          1217   NaN     RAMIREZ   
87250              7.0  ...  2º semestre          1973   NaN       LOPEZ   

          nombre id_puesto                               email  \
1073       MARIA  70-21120           maria.amaya@tractchun.com   
91273  GUADALUPE  15-24100  guadalupe.montenegro@tractchun.com   
44316  DOMINGUEZ  80-22210     dominguez.ramirez@tractchun.com   
87250   DIOGENES  80-22200        diogenes.lopez@tractchun.com   

                                    puesto      departamento     tipo_puesto  
1073          Auxiliar de Recursos Humanos  Recursos Humanos  Administrativo  
91273             Supervisor de Producción        PRODUCCION   Mandos Medios  
44316          Supervisor de Mantenimiento     MANTENIMIENTO   Mandos Medios  
87250  Supervisor General de Mantenimiento     MANTENIMIENTO   Mandos Medios  

[4 rows x 22 columns]

In [38]:
df_sat_emp = df_completo.groupby("id_empleado")["satisfacción"].mean().reset_index()
df_sat_emp = df_sat_emp.sort_values(by = "satisfacción")

In [39]:
df_sat_emp

id_empleado  satisfacción
838         839.0      3.488372
1724       1725.0      3.567568
1824       1825.0      3.571429
1998       1999.0      3.575000
1365       1366.0      3.577778
...           ...           ...
1835       1836.0      4.586957
1942       1943.0      4.589286
858         859.0      4.612245
745         746.0      4.641026
621         622.0      4.744681

[2000 rows x 2 columns]

In [40]:
print("Los empleados con más satisfacción media son: ")
df_sat_emp.tail()

Los empleados con más satisfacción media son: 


id_empleado  satisfacción
1835       1836.0      4.586957
1942       1943.0      4.589286
858         859.0      4.612245
745         746.0      4.641026
621         622.0      4.744681

In [41]:
print("Los empleados con menos satisfacción media son: ")
df_sat_emp.head()

Los empleados con menos satisfacción media son: 


id_empleado  satisfacción
838         839.0      3.488372
1724       1725.0      3.567568
1824       1825.0      3.571429
1998       1999.0      3.575000
1365       1366.0      3.577778